src/logic/__tests__/validateCurrentProject.test.js

```diff
@@ -0,0 +1,56 @@
+import validateCurrentProject from '../validateCurrentProject';
+
+import Analyzer from '../../analyzers';
+import validateSource from '../helpers/validateSource';
+import {getCurrentProject} from '../../selectors';
+
+const mockHtmlSource = '<html></html>';
+const mockCssSource = 'div {}';
+
+jest.mock('../../analyzers');
+jest.mock('../helpers/validateSource');
+jest.mock('../../selectors', () => ({
+  getCurrentProject: jest.fn(() => ({
+    sources: {
+      html: mockHtmlSource,
+      css: mockCssSource,
+    },
+  })),
+}));
+
+const dummyState = {};
+
+test('should validate current project', async () => {
+  const dispatch = jest.fn();
+  const done = jest.fn();
+  await validateCurrentProject.process(
```

Comment:
So, this is a bit of a pain in the ass, but `await`ing the call to `process` is actually testing implementation details of that method; from the standpoint of `redux-logic`, `process` does not have to be an async (or Promise-returning) function; it just needs to call `done()` when it’s done. I think the best way to test the relevant interface is something like:

```js
await new Promise((resolve) => {
  validateCurrentProject.process(
    {getState: constant(dummyState)},
    dispatch,
    resolve,
  )
});
```

FWIW I was looking through other logic tests for examples and am realizing I totally failed to catch a bunch of other cases of what you did here, so I’ll need to go back and fix those—sorry!!

Reply:
Hey so! I dug into fixing this issue for the existing tests, and ended up just writing a test helper that actually runs logics through the `redux-logic` middleware, so the tests can focus on what the logics _do_ rather than trying to mimic the implementation of `redux-logic`. Check out #2026 for all the sordid details!